lib/devise/migratable/model.rb

```diff
@@ -0,0 +1,146 @@
+    #
+    # This module requires a new database column `encrypted_password_migrate_to` to store the new hashed
+    # password. So for each password set or validate, it would start storing the new hashed password
+    # in encrypted_password_migrate_to. If feature flag enabled, it would start validating against the
+    # encrypted_password_migrate_to. The goal of this module is to allow us gracefully migrate existing password
+    # hashing with the ability to rollback or turn on and off through feature flags.
+    #
+    # Once all users have passwords hashed in the encrypted_password_migrate_to, you could run a migration
+    # to copy/swap data from encrypted_password_migrate_to to encrypted_password, at the same time, replace
+    # this module with `devise-encrytable`.
+    #
+    # == Options
+    #
+    # Migratable adds the following options to devise_for:
+    #
+    #   * +pepper+: a random string used to provide a more secure hash.
+    #
+    #   * +encryptor+: the encryptor going to be used. By default is nil.
+    #
+    #   * +feature_class+: feature class used to check wether validating password using new encryptor.
+    #
+    #   * +feature_name+: feature name will be passed to feature_class.active?(feature_name, user_model).
+    #
+    # == Examples
+    #    # with feature disabled
+    #    User.find(1).valid_password?('password123')
+    #    #=> returns true/false
+    #    #=> if true will start saving new hashed password to encrypted_password_migrate_to using encryptor
+    #    #=> still using encrypted_password to validate because feature not enabled
+    #
+    #    # with feature enabled
+    #    User.find(1).valid_password?('password123')
+    #    #=> returns true/false
+    #    #=> validate against encrypted_password_migrate_to using encryptor, fallback to old one if encrypted_password_migrate_to not present
+    module Migratable
+      extend ActiveSupport::Concern
+
+      def self.required_fields(_klass)
+        [:encrypted_password_migrate_to]
+      end
+
+      # Generates new password for the new encryptor
+      # generate password of the old one using super
+      def password=(new_password)
+        # everytime we store the new one with the given encryptor
+        self.encrypted_password_migrate_to = generate_digest_for_password(new_password)
+        # this will set the original password to the database, nothing changed
+        super
+      end
+
+      # Validates the password considering the salt.
+      def valid_password?(password)
+        return false if encrypted_password.blank?
+
+        # only if feature enabled? then we do the work here
+        if feature_enabled? && encrypted_password_migrate_to.present?
+          valid_password_using_encryptor?(password)
+        else
+          result = super
+          if result && encrypted_password_migrate_to.nil?
+            update_attribute(:encrypted_password_migrate_to, generate_digest_for_password(password))
+          end
```

Comment:
Make sense, I wrapped `update_attribute` call with rescue, rather than entire method using `ensure` cause I still want to return the new validation result if the feature flag is on, let me know if you have other concerns.